internal/docstore/firedocstore/fs.go

```diff
@@ -586,6 +588,65 @@ func processMods(mods []driver.Mod) (fields map[string]*pb.Value, maskPaths []st
 	return fields, maskPaths, transforms, nil
 }
 
+// doCommitCall Calls the Commit RPC with a list of writes, and handles the results.
+func (c *collection) doCommitCall(ctx context.Context, call *commitCall, errs []error, opts *driver.RunActionsOptions) {
+	wrs, err := c.commit(ctx, call.writes, opts)
+	if err != nil {
+		for _, a := range call.actions {
+			errs[a.Index] = err
+		}
+		return
+	}
+	_ = wrs
+	// Set the revision fields of the documents.
+	// The actions and writes may not correspond, because Update actions may require
+	// two writes. We can tell which writes correspond to actions by the type of write.
+	// TODO(jba): enable this in a separate PR; it breaks tests.
+	// TODO(jba): test that writes return revisions in drivertest.go.
+	// j := 0
+	// for i, wr := range wrs {
+	// 	if _, ok := call.writes[i].Operation.(*pb.Write_Transform); !ok {
+	// 		// Ignore errors. It's fine if the doc doesn't have a revision field.
+	// 		_ = call.actions[j].Doc.SetField(docstore.RevisionField, wr.UpdateTime)
+	// 		j++
+	// 	}
+	// }
+	// Set new names for Create actions.
+	for i, a := range call.actions {
+		if call.newNames[i] != "" {
+			_ = a.Doc.SetField(c.nameField, call.newNames[i])
+		}
+	}
+}
+
+func (c *collection) commit(ctx context.Context, ws []*pb.Write, opts *driver.RunActionsOptions) ([]*pb.WriteResult, error) {
+	req := &pb.CommitRequest{
+		Database: c.dbPath,
+		Writes:   ws,
+	}
+	if opts.BeforeDo != nil {
+		asFunc := func(i interface{}) bool {
+			p, ok := i.(**pb.CommitRequest)
+			if !ok {
+				return false
+			}
+			*p = req
+			return true
+		}
+		if err := opts.BeforeDo(asFunc); err != nil {
+			return nil, err
+		}
+	}
+	res, err := c.client.Commit(withResourceHeader(ctx, req.Database), req)
+	if err != nil {
+		return nil, err
+	}
+	if len(res.WriteResults) != len(ws) {
+		return nil, gcerr.Newf(gcerr.Internal, nil, "wrong number of WriteResults from firestore commit")
```

Comment:
This error will be confusing when the `doCommitCall` method assign it to the no-precondition calls, as those writes can be separate. I don't think in practice this is a big issue but since we are making many assumptions on how the user is using the API, maybe we should 1) document the behavior; 2) provide a way to still let user run things in batches when they need that.

Reply:
Really, this is an impossible error. It would mean that firestore is very broken. I only added this so that we don't panic. I think we can basically ignore it.